aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,406 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Rules for the AiiDA Graph Explorer utility"""
+
+from abc import ABCMeta, abstractmethod
+from collections import defaultdict
+import numpy as np
+
+from aiida.common.exceptions import InputValidationError
+from aiida.orm.querybuilder import QueryBuilder
+from aiida.tools.graph.age_entities import Basket
+
+
+def check_if_basket(entity_set):
+    if not isinstance(entity_set, Basket):
+        raise TypeError('You need to set the walkers with an AiidaEntitySet')
+
+
+class Operation(metaclass=ABCMeta):
+    """Operation metaclass - Base class for all AGE explorer classes"""
+
+    def __init__(self, max_iterations, track_edges):
+        self.set_max_iterations(max_iterations)
+        self._track_edges = track_edges  # Logical
+        self._iterations_done = None
+
+    def set_max_iterations(self, max_iterations):
+        """Set max iterations"""
+        if max_iterations is np.inf:
+            pass
+        elif not isinstance(max_iterations, int):
+            raise TypeError('max iterations has to be an integer')
```

Comment:
simplify into
```
if not (isinstance(max_iterations, int) or max_iterations is inf):
```
and adapt exception message to say it allows `np.inf`

Reply:
Is using an intermediate variable also acceptable or you still would prefer the one line `if`?

```
valid_input = isinstance(max_iterations, int) or max_iterations is np.inf
if not valid_input:
    (...)
```

Reply:
What's the reason? If the exact line is used multiple times I see the point, but if you use it just in the conditional I think it is better to keep the single line.

Reply:
I don't know, it feels more "self-documented" and faster to understand if have one line that says "this is a valid input if these conditions are met" and then the actual conditional "if this is a valid input, then..." than to just putting all inside the conditional block and having to analyze the conditions being checked + conditional branches to understand what is happening.

In this case the effect is immediate by reading the error raised and the conditions are not that hard to understand, and so `valid_input` is not adding that much information, so I'm going to use the one line. But as a general rule, I kind of like the approach of "collapsing" checks in adequately named booleans and just checking those when using conditional blocks when there are more than one "direct check" to make. Do you see any drawback to this / don't agree that having the "intention" stating variable adds clarity to the code?